### PR TITLE
[k8s] Add option to divide ratelimit by number of kubernetes pods

### PR DESCRIPTION
**Pull request by cbrewster**
We wanted a way to divide the ratelimits by the number of serving pods,
this adds a way to determine the pod count dynamically and divides the
limits by the pod count.

### CODE REVIEW

**Comment by cbrewster** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1** <a href="https://app.graphite.dev/github/pr/replit/rrl/1" target="_blank">Graphite</a> 
    * **PR #2** <a href="https://app.graphite.dev/github/pr/replit/rrl/2" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/replit/rrl/2?utm_source=stack-comment).

File: plugins/rrl/handler.go [hunk 1]
```diff
@@ -26,11 +27,14 @@ func (rrl *RRL) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 
 	// Limit request rate
 	if rrl.requestsInterval != 0 {
+		// We multuply because allowance is (1s / RPS)
```
Comment (ryantm) [inline plugins/rrl/handler.go]:
```suggestion
		// We multiply because allowance is (1s / RPS)
```

File: plugins/rrl/handler.go [hunk 2]
```diff
@@ -26,11 +27,14 @@ func (rrl *RRL) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
...
+		requestsInterval := rrl.requestsInterval * atomic.LoadInt64(&rrl.podCount)
```
Comment (ryantm) [inline plugins/rrl/handler.go:31]: Do we need some minimum to make sure the request limit doesn't get too low?

File: plugins/rrl/rrl.go [hunk 3]
```diff
@@ -77,19 +80,39 @@ func responseType(m *dns.Msg) byte {
...
-		return rrl.referralsInterval
+		allowance = rrl.referralsInterval
 	case rTypeError:
-		return rrl.errorsInterval
+		allowance = rrl.errorsInterval
+	default:
+		return -1
+	}
+
+	// We multuply because allowance is (1s / RPS)
```
Comment (ryantm) [inline plugins/rrl/rrl.go]:
```suggestion
	// We multiply because allowance is (1s / RPS)
```

**Review by ryantm — APPROVED** [review]

File: plugins/rrl/rrl.go (hunk 3 above)
Comment (lhchavez) [inline plugins/rrl/rrl.go]:
```suggestion
	// We multiply because allowance is (1s / RPS)
```

File: plugins/rrl/rrl.go [hunk 4]
```diff
@@ -77,19 +80,39 @@ func responseType(m *dns.Msg) byte {
...
+	return allowance * atomic.LoadInt64(&rrl.podCount)
+}
+
+// scanKubernetesPods periodically determines the number of pods runnings in kubernetes
```
Comment (lhchavez) [inline plugins/rrl/rrl.go]:
```suggestion
// scanKubernetesPods periodically determines the number of pods running in kubernetes
```

File: plugins/rrl/handler.go (hunk 2 above)
Comment (cbrewster) [reply plugins/rrl/handler.go:31]: Ah yeah we should min pod count at 1

**Review by lhchavez — APPROVED** [review]

File: plugins/rrl/handler.go (hunk 2 above)
Comment (ryantm) [reply plugins/rrl/handler.go:31]: That sounds good. But I said the wrong thing in my above comment. What I meant was, what if the requestsInterval gets so high that the individual pods effectively serve no traffic?

Comment (cbrewster) [reply plugins/rrl/handler.go:31]: We do have a maximum number of pods we allow the deployment to scale to, we should be able to account for that when selecting our new ratelimit values.

**Pull request merged**